File: Chapter03(Fingerprint)/FingerPrintReader/FingerPrintReader/DBClass.vb
'-------------------------------------------------------------------------------
'GrFinger Sample
'(c) 2005 Griaule Tecnologia Ltda.
'http://www.griaule.com
'-------------------------------------------------------------------------------
'
'This sample is provided with "GrFinger Fingerprint Recognition Library" and
'can't run without it. It's provided just as an example of using GrFinger
'Fingerprint Recognition Library and should not be used as basis for any
'commercial product.
'
'Griaule Tecnologia makes no representations concerning either the merchantability
'of this software or the suitability of this sample for any particular purpose.
'
'THIS SAMPLE IS PROVIDED BY THE AUTHOR "AS IS" AND ANY EXPRESS OR
'IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
'OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
'IN NO EVENT SHALL GRIAULE BE LIABLE FOR ANY DIRECT, INDIRECT,
'INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
'NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
'DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
'THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
'(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
'THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
'
'You can download the free version of GrFinger directly from Griaule website.
'
'These notices must be retained in any copies of any part of this
'documentation and/or sample.
'
'-------------------------------------------------------------------------------

' -----------------------------------------------------------------------------------
' Database routines
' -----------------------------------------------------------------------------------

Imports System.Data.OleDb
Imports System.Runtime.InteropServices

' Template data
Public Class TTemplate
	' Template itself
    Public tpt As System.Array = Array.CreateInstance(GetType(Byte), GrFingerXLib.GRConstants.GR_MAX_SIZE_TEMPLATE)


	' Template size
    Public Size As Long
End Class

' Template list
Public Structure TTemplates
	' ID
	Public ID As Integer
	' Template itself
	Public template As TTemplate
End Structure

Public Class DBClass

	' the database we'll be connecting to
    Const DBFile As String = "GrFingerSample.mdb"
    Const ConnectionString As String = "Provider=Microsoft.Jet.OLEDB.4.0;Data Source="

	' the connection object
	Dim connection As New OleDbConnection

	' Open connection
	Public Function OpenDB() As Boolean
		Dim filePath As String
		Try
			filePath = Application.StartupPath() & "\" & DBFile
			connection = New OleDb.OleDbConnection(ConnectionString & filePath)
			Return True
		Catch
			Return False
		End Try
	End Function

	' Close conection
	Public Sub closeDB()
		connection.Close()
	End Sub

	' Clear database
	Public Sub clearDB()
		Dim sqlCMD As OleDbCommand = New OleDbCommand("DELETE FROM enroll", connection)
		' run "clear" query
		sqlCMD.Connection.Open()
		sqlCMD.ExecuteNonQuery()
		sqlCMD.Connection.Close()
	End Sub

	' Add template to database. Returns added template ID.
	Public Function AddTemplate(ByRef template As TTemplate) As Long
		Dim da As New OleDbDataAdapter("select * from enroll", connection)

		' Create SQL command containing ? parameter for BLOB.
		da.InsertCommand = New OleDbCommand("INSERT INTO enroll (template) Values(?)", connection)
		da.InsertCommand.CommandType = CommandType.Text
		da.InsertCommand.Parameters.Add("@template", OleDbType.Binary, template.Size, "template")

		' Open connection
		connection.Open()

		' Fill DataSet.
		Dim enroll As DataSet = New DataSet
		da.Fill(enroll, "enroll")

		' Add a new row.
        ' Create parameter for ? contained in the SQL statement.
		Dim newRow As DataRow = enroll.Tables("enroll").NewRow()
		newRow("template") = template.tpt
		enroll.Tables("enroll").Rows.Add(newRow)

		' Include an event to fill in the Autonumber value.
		AddHandler da.RowUpdated, New OleDbRowUpdatedEventHandler(AddressOf OnRowUpdated)

		' Update DataSet.
		da.Update(enroll, "enroll")
		connection.Close()

		' return ID
		Return newRow("ID")
	End Function

	' Event procedure for OnRowUpdated
	Private Sub OnRowUpdated(ByVal sender As Object, ByVal args As OleDbRowUpdatedEventArgs)
		' Include a variable and a command to retrieve identity value
		' from Access database.
		Dim newID As Integer = 0
		Dim idCMD As OleDbCommand = New OleDbCommand("SELECT @@IDENTITY", connection)

		If args.StatementType = StatementType.Insert Then
			' Retrieve identity value and store it in column
			newID = CInt(idCMD.ExecuteScalar())
			args.Row("ID") = newID
		End If
	End Sub

	' Returns a DataTable with all enrolled templates from database.
	Public Function getTemplates() As TTemplates()
		Dim ds As New DataSet
		Dim da As New OleDbDataAdapter("select * from enroll", connection)
		Dim ttpts As TTemplates()
		Dim i As Integer

		' Get query response
		da.Fill(ds)
		Dim tpts As DataRowCollection = ds.Tables(0).Rows
		' Create response array
		ReDim ttpts(tpts.Count)
		' No results?
		If tpts.Count = 0 Then Return ttpts
		' get each template and put results in our array
		For i = 1 To tpts.Count
			ttpts(i).template = New TTemplate
			ttpts(i).ID = tpts.Item(i - 1).Item("ID")
			ttpts(i).template.tpt = tpts.Item(i - 1).Item("template")
			ttpts(i).template.Size = ttpts(i).template.tpt.Length
		Next
		Return ttpts
	End Function

	' Returns template with supplied ID.
	Public Function getTemplate(ByVal id As Long) As Byte()
        Dim ds As New DataSet
		Dim da As New OleDbDataAdapter("select * from enroll where ID = " & id, connection)
		Dim tpt As New TTemplate

		' Get query response
		da.Fill(ds)
		Dim tpts As DataRowCollection = ds.Tables(0).Rows
		' No results?
        If tpts.Count <> 1 Then Return Nothing
        ' Deserialize template and return it
        Return tpts.Item(0).Item("template")
	End Function

End Class

File: Chapter03(Fingerprint)/FingerPrintReader_CS/FingerPrintReader_CS/DBClass.cs
/*
 -------------------------------------------------------------------------------
 GrFinger Sample
 (c) 2005 Griaule Tecnologia Ltda.
 http://www.griaule.com
 -------------------------------------------------------------------------------

 This sample is provided with "GrFinger Fingerprint Recognition Library" and
 can't run without it. It's provided just as an example of using GrFinger
 Fingerprint Recognition Library and should not be used as basis for any
 commercial product.

 Griaule Tecnologia makes no representations concerning either the merchantability
 of this software or the suitability of this sample for any particular purpose.

 THIS SAMPLE IS PROVIDED BY THE AUTHOR "AS IS" AND ANY EXPRESS OR
 IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 IN NO EVENT SHALL GRIAULE BE LIABLE FOR ANY DIRECT, INDIRECT,
 INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

 You can download the free version of GrFinger directly from Griaule website.
                                                                   
 These notices must be retained in any copies of any part of this
 documentation and/or sample.

 -------------------------------------------------------------------------------
*/

// -----------------------------------------------------------------------------------
// Database routines
// -----------------------------------------------------------------------------------

using System;
using System.Data;
using System.Data.OleDb;
using GrFingerXLib;
using System.Runtime.InteropServices;

// the template class
public class TTemplate
{
	// Template data.
	public System.Array _tpt;
	// Template size
	public int _size;

	public TTemplate(){
		// Create a byte buffer for the template
	   _tpt = new byte[(int)GRConstants.GR_MAX_SIZE_TEMPLATE];
	   _size = 0;
	}
}

// the database class
public class DBClass{
	
	// the connection object
	private OleDbConnection _connection;

	// temporary template for retrieving data from DB
	private TTemplate tptBlob;
	
	// the database we'll be connecting to
	public readonly string CONNECTION_STRING = "Provider=Microsoft.Jet.OLEDB.4.0;Data Source=GrFingerSample.mdb";
	
	public DBClass(){
	}

	// Open connection
	public bool openDB()
	{
		_connection = new OleDbConnection();
		_connection.ConnectionString = CONNECTION_STRING;
		try{
			_connection.Open();
		}
		catch{
			return false;
		}
		tptBlob = new TTemplate();
		return true;
	}//END

	// Close conection
	public bool closeDB()
	{
		if(_connection.State != ConnectionState.Closed)
		  _connection.Close();  
		return true;
	}

	// Clear database
	public bool clearDB()
	{
		OleDbCommand cmdClear = null;
		cmdClear = new OleDbCommand("DELETE FROM enroll", _connection);

		// run "clear" query
		if(_connection.State == ConnectionState.Open)
			cmdClear.ExecuteNonQuery();
		
		return true;
	}


	// Add template to database. Returns added template ID.
	public bool addTemplate(TTemplate tpt,ref int id) 
	{
		OleDbCommand cmdInsert = null;
		OleDbParameter dbParamInsert = null; 
		OleDbCommand cmdSelect =  null;

		try{
			// Create SQL command containing ? parameter for BLOB.
			cmdInsert = new OleDbCommand("INSERT INTO enroll(template) values(?) ", _connection);
			// Create parameter for ? contained in the SQL statement.
			System.Byte [] temp = new System.Byte[tpt._size + 1];
			System.Array.Copy(tpt._tpt, 0, temp, 0, tpt._size);

			dbParamInsert = new OleDbParameter("@template", OleDbType.VarBinary, tpt._size, 
							ParameterDirection.Input, false, 0, 0,"ID", 
							DataRowVersion.Current, temp);
			cmdInsert.Parameters.Add(dbParamInsert);

			//execute query
			if(_connection.State == ConnectionState.Open)
				cmdInsert.ExecuteNonQuery();
		}
		catch{
			return false;
		}

		try{
			// Create SQL command containing ? parameter for BLOB.
			cmdSelect = new OleDbCommand("SELECT top 1 ID FROM enroll ORDER BY ID DESC", _connection);
		    
			id = System.Convert.ToInt32(cmdSelect.ExecuteScalar());
		}
		catch {
			return false;  
		}

		return true;
	}

	// Returns an OleDbDataReader with all enrolled templates from database.
	public OleDbDataReader  getTemplates()
	{
		OleDbCommand  cmdGetTemplates;
		OleDbDataReader  rs;

		//setting up command 
		cmdGetTemplates =  new OleDbCommand("SELECT * FROM enroll", _connection);
		rs = cmdGetTemplates.ExecuteReader();

		return rs;
	}

	
	// Returns template with the supplied ID.
	public TTemplate getTemplate(int id)
	{
		OleDbCommand cmd = null;
		OleDbDataReader dr = null;
		tptBlob._size = 0;
		try
		{
			cmd =  new OleDbCommand(System.String.Concat("SELECT * FROM enroll WHERE ID = ", System.Convert.ToString((int)id)), _connection);
			dr = cmd.ExecuteReader();
			// Get query response
			dr.Read();
			getTemplate(dr);
			dr.Close();
		}
		catch{
			dr.Close();
		}
		return tptBlob;
	}

	// Return template data from an OleDbDataReader
	public TTemplate getTemplate(OleDbDataReader rs)
	{
		long readedBytes; 
		tptBlob._size = 0;
		// alloc space
		System.Byte[] temp = new System.Byte[
			(int)GRConstants.GR_MAX_SIZE_TEMPLATE];
		// get bytes
		readedBytes = rs.GetBytes(1, 0, temp, 0,temp.Length);
		// copy to structure
		System.Array.Copy(temp, 0, tptBlob._tpt,0,(int)readedBytes);
		// set real size
		tptBlob._size = (int)readedBytes;

		return tptBlob;
	}

	// Return enrollment ID from an OleDbDataReader
	public int getId(OleDbDataReader rs)
	{
		return rs.GetInt32(0);
	}
}
